### ISSUE TITLE
Memory leak from repeatedly pressing load default palette

### ISSUE THREAD

**Issue by Kemosalamy** [synfig/synfig#2598]
**Describe the bug**
Memory of the synfig application increases as the load default palette button is pressed repeatedly.

**To Reproduce**
1. Open synfig
2. Repeatedly press the load default palette button
3. View program memory usage(keeps increasing)

**Expected behavior**
Memory usage does not keep increasing as the button is repeatedly pressed.

**Screenshots**
Button to press:
![image](https://user-images.githubusercontent.com/81593733/161317989-87b7f4d1-f4aa-4702-95e9-92380cae2fd5.png)

Memory before pressing:
![image](https://user-images.githubusercontent.com/81593733/161318045-3526f314-deee-40b8-99ae-3da8cef2207c.png)

Memory after pressing repeatedly for a 30 seconds:
![image](https://user-images.githubusercontent.com/81593733/161318503-08e61ee1-3821-45ae-9658-6f0c45671e2f.png)

**System information:**
 - OS: Windows
 - OS version: Windows 10
 - Synfig master branch
  

**Additional comments**
The refresh() function alone produces the behaviour described in this issue.

**Comment by ice0** [comment]
@Kemosalamy
Do you plan to fix this?

**Comment by Kemosalamy** [comment]
@ice0 Yes I will start investigating this problem right away, please assign me to this issue 👍

**Comment by Kemosalamy** [comment]
One of the issues seems to have been specifically with this part of the code:
https://github.com/synfig/synfig/blob/9b5140ffb49b6ec64a4c729e8862b9c87cb791e0/synfig-studio/src/gui/modules/mod_palette/dock_paledit.cpp#L374-L385

Specifically in line 384 where new Widget_Color* s are created everytime the table is refreshed. the problem is that when the refresh function is called again, the previous Widget_Color* s are removed in the line ```table.foreach(sigc::mem_fun(table,&Gtk::Table::remove));```  however their memory is not freed. 

I have added a draft pull request https://github.com/synfig/synfig/pull/2633 which solves this issue by looping over the children before they are removed and freeing their memory using delete. This was the cause of the most major memory leak however there still seems to be one or two other minor memory leaks. I will investigate them further.

**Comment by ice0** [comment]
> `table.foreach(sigc::mem_fun(table,&Gtk::Table::remove));` however their memory is not freed.

Yes, because it is removed from the table, but not destroyed. This is expected behavior.
https://docs.gtk.org/gtk3/method.Container.remove.html

> If you don’t want to use widget again it’s usually more efficient to simply destroy it directly using gtk_widget_destroy() since this will remove it from the container and help break any circular reference count cycles.

Why don't you use `table.foreach(...)` for this?

**Comment by Kemosalamy** [comment]
It seems the gtk_widget_destroy() function does not exactly work with the children object type of table(table has children of type Widget* whereas gtk_widget_destroy() take as an input a GtkWidget*) so i'm not sure how i would typecast in a foreach function. I'm also confused as to which documentation I should be following as most of the code seems to be using the gtkmm classes and libraries(however gtkmm does not seem to have a gtk_widget_destroy() or at least I was unable to find it) but the documentation for the Container.remove was from a Gtk library.

**Comment by ice0** [comment]
This is gtkmm docs for Container:

https://developer-old.gnome.org/gtkmm/3.24/classGtk_1_1Container.html#a9f31c07118f7bdc7a4e0651acf35abbc

> If widget is managed with [Gtk::manage()](https://developer-old.gnome.org/gtkmm/3.24/namespaceGtk.html#a753d0ddca83bd0d0decdfdcd1e93076a), and you don't want to use widget again, then you should delete it, because there will no longer be any parent container to delete it automatically.
